src/plonk/lookup/prover.rs

```diff
@@ -327,35 +330,44 @@ impl<C: CurveAffine> Permuted<C> {
         // Compute the evaluations of the lookup product polynomial
         // over our domain, starting with z[0] = 1
         let z = iter::once(C::Scalar::one())
-            .chain(lookup_product.into_iter().skip(1))
+            .chain(lookup_product)
             .scan(C::Scalar::one(), |state, cur| {
                 *state *= &cur;
                 Some(*state)
             })
+            // Take all rows including the "last" row which should
+            // be a boolean (and ideally 1, else soundness is broken)
+            .take(params.n as usize - blinding_factors)
+            // Chain random blinding factors.
+            .chain((0..blinding_factors).map(|_| C::Scalar::rand()))
             .collect::<Vec<_>>();
+        assert_eq!(z.len(), params.n as usize);
         let z = pk.vk.domain.lagrange_from_vec(z);
 
         #[cfg(feature = "sanity-checks")]
         // This test works only with intermediate representations in this method.
         // It can be used for debugging purposes.
         {
             // While in Lagrange basis, check that product is correctly constructed
-            let n = params.n as usize;
+            let n = (params.n as usize) - (blinding_factors + 1);
```

Comment:
We call this u in https://github.com/zcash/halo2/blob/750ab469d0d71090ce56834fc78e46db25f5250b/book/src/design/proving-system/lookup.md#zero-knowledge-adjustment .